File: Coursera - Computational Geometry/PointInclusionInConvexPolygon/PointInclusionInConvexPolygon.js
const eps = 0.00001;
class Point2D {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }
}

//return < 0 if tg1 < tg2
function comparePoints(p1, p2, M) {
  return function(p1, p2) {
    const tg1 = (Math.abs(p1.x - M.x) > eps) ? (p1.y - M.y) / (p1.x - M.x) : null;
    const tg2 = (Math.abs(p2.x - M.x) > eps) ? (p2.y - M.y) / (p2.x - M.x) : null;

    if(!tg1 || !tg2) {
      if(Math.abs(p1.x - p2.x) > eps) {
        return p1.x - p2.x;
      } else {
        return p1.y - p2.y;
      }
    }

    if(Math.abs(tg1 - tg2) < eps) {
      return p1.y - p2.y;
    }

    return tg1 - tg2;

  }
}

//Test the compare function
let p1, p2, M;

M = new Point2D(0, -3);
p1 = new Point2D(2, 1);
p2 = new Point2D(-2, 1);

console.log(comparePoints(p1, p2, M)());

function isPointInsidePolygon(p, polygon) {
  const M = new Point2D(
    (polygon[0].x + polygon[1].x + polygon[2].x) / 3,
    (polygon[0].y + polygon[1].y + polygon[2].y) / 3
  );
  const cmp = comparePoints(polygon[0], polygon[1], M);
  cmp(polygon[0], polygon[1]);
  console.log(M);
}

//Test 1
const polygon = [
  new Point2D(-2, -3),
  new Point2D(1, -4),
  new Point2D(3, -2),
  new Point2D(2, 1),
  new Point2D(-2, 1),
];

const points = [
  new Point2D(2, -3),
  new Point2D(3, 0),
  new Point2D(0, 0),
  new Point2D(2, -1),
];

isPointInsidePolygon(points[0], polygon);

File: Coursera - Computational Geometry/PointInclusionInConvexPolygon/PointInclusionInConvexPolygon.py
import unittest
import math

EPS = 0.000001

def eq(x, y):
  return abs(x - y) < EPS

def neq(x, y):
  return not eq(x, y)

class Point:
  def __init__(self, x, y):
    self.x = x
    self.y = y
  
  def __str__(self):
    return "(" + str(self.x) + " ," + str(self.y) + ")"

# If p1 < p2 return something negative. Else return something positive.
def createCmp(M):
  def cmp(p1, p2):
    d = det(M, p2, p1)
    return d if d != 0 else p1.x - p2.x 

  return cmp

def det(a, b, c):
  return (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x)


def computeCenterOfMass(polygon):
  return Point((polygon[0].x + polygon[1].x + polygon[2].x) / 3.0, (polygon[0].y + polygon[1].y + polygon[2].y) / 3.0)

# Driver function to solve the problem
def solve(polygon, points):
  M = computeCenterOfMass(polygon)
  

# Determine in which wedge the point p lies.
# pivot is a point from inside the polygon
# polygon represents an array of verticies in counterclockwise order
# Wedge n is given by polygon[n] and polygon[n + 1]
def findWedge(polygon, pivot, p):
  left = 0
  right = len(polygon) - 1
  m = 0
  solution = 0
  while left <= right:
    m = math.floor((left + right) / 2)
    d = det(pivot, polygon[m], p)

    if d == 0:
      break
    if d < 0:
      right = m - 1
      solution = m - 1
      if solution < 0:
        solution = len(polygon) - 1
    else:
      left = m + 1
      solution = m
    

  return solution

#Test 1

polygon = [
  Point(-2, -3),
  Point(1, -4),
  Point(3, -2),
  Point(2, 1),
  Point(-2, 1)
]

points = [
  Point(2, -3),
  Point(3, 0),
  Point(0, 0),
  Point(2, -1)
]

#solve(polygon, points)
